Installation issue: esmf

### Steps to reproduce the issue

ESMF fails to compile on NOAA WCOSS2/Acorn systems with Cray PE+Intel compilers+Cray MPICH. Builds successfully if ESMF_OS=Linux and ESMF_COMM=mpich3, but esmf/package.py sets those to Unicos and mpi, respectively.

```console
$ spack spec -l esmf
Input spec
--------------------------------
esmf

Concretized
--------------------------------
6rxazwx  esmf@8.3.0b09%intel@19.1.3.304~debug~external-lapack+mpi+netcdf~parallelio+pio~pnetcdf~shared~xerces build_system=makefile arch=linux-sles15-zen2
zhz3fn4      ^cray-mpich@8.1.9%intel@19.1.3.304~wrappers build_system=generic arch=linux-sles15-zen2
7vjuutq      ^libxml2@2.10.3%intel@19.1.3.304~python build_system=autotools arch=linux-sles15-zen2
inla2gw          ^libiconv@1.16%intel@19.1.3.304 build_system=autotools libs=shared,static arch=linux-sles15-zen2
eyz7fdy          ^pkg-config@0.29.2%intel@19.1.3.304+internal_glib build_system=autotools arch=linux-sles15-zen2
ilrrzhx          ^xz@5.2.6%intel@19.1.3.304~pic build_system=autotools libs=shared,static arch=linux-sles15-zen2
e65xq45      ^netcdf-c@4.7.4%intel@19.1.3.304~dap~fsync~hdf4~jna+mpi+optimize~parallel-netcdf+pic~shared build_system=autotools arch=linux-sles15-zen2
xc2opw5          ^hdf5@1.10.6%intel@19.1.3.304~cxx+fortran+hl~ipo~java+mpi~shared~szip+threadsafe+tools api=default build_system=cmake build_type=RelWithDebInfo arch=linux-sles15-zen2
utbfs5w              ^cmake@3.20.2%intel@19.1.3.304~doc+ncurses+ownlibs~qt build_system=generic build_type=Release arch=linux-sles15-zen2
sj5fkki          ^m4@1.4.18%intel@19.1.3.304+sigsegv build_system=autotools patches=3877ab5,fc9b616 arch=linux-sles15-zen2
cry7dbu      ^netcdf-fortran@4.5.4%intel@19.1.3.304~doc+pic~shared build_system=autotools arch=linux-sles15-zen2
ij722et      ^zlib@1.2.11%intel@19.1.3.304+optimize+pic~shared build_system=makefile arch=linux-sles15-zen2
...
```

@climbfuji @jedwards4b

### Error message

Here is a sampling of the error messages (other undefined references include mpi_send_, mpi_wait_, etc.):
<details><summary>Error message</summary><pre>
  >> 13388    /usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/cache/build_stage/spack-stage-esmf-8.3.0b09-jmblchtb6cpbt4ot6f22x45qwyrju334/spack-src/src/Infrastr
              ucture/IO/PIO/piodarray.F90.in:162: undefined reference to `mpi_bcast_'
  >> 13389    /usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/cache/build_stage/spack-stage-esmf-8.3.0b09-jmblchtb6cpbt4ot6f22x45qwyrju334/spack-src/src/Infrastr
              ucture/IO/PIO/piodarray.F90.in:163: undefined reference to `mpi_bcast_'
  >> 13390    /usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/cache/build_stage/spack-stage-esmf-8.3.0b09-jmblchtb6cpbt4ot6f22x45qwyrju334/spack-src/src/Infrastr
              ucture/IO/PIO/piodarray.F90.in:164: undefined reference to `mpi_bcast_'
  >> 13391    /usr/lib64/gcc/x86_64-suse-linux/7/../../../../x86_64-suse-linux/bin/ld: /path/to/cache/build_stage/spack-stage-esmf-8.3.0b09-jmblchtb6cpbt4ot6f22x45qwyrju334/spack-src/src/Infrastr
              ucture/IO/PIO/piodarray.F90.in:165: undefined reference to `mpi_bcast_'
</pre></details>


### Information on your system

Lmod modules in ESMF build env:
1) craype-x86-rome (H)   2) envvar/1.0   3) PrgEnv-intel/8.3.3   4) intel/19.1.3.304   5) craype/2.7.13   6) libfabric/1.11.0.0. (H)   7) craype-network-ofi (H)   8) cray-mpich/8.1.9

spack debug report:
* **Spack:** 0.20.0.dev0 (3a152138df8f6db58019b1b8d19b75a0bbbd0c23)
* **Python:** 3.6.15
* **Platform:** linux-sles15-zen2
* **Concretizer:** clingo

### Additional information

[spack-build-env.txt](https://github.com/spack/spack/files/10318160/spack-build-env.txt)
[spack-build-out.txt](https://github.com/spack/spack/files/10318162/spack-build-out.txt)


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/esmf/package.py
+++ b/var/spack/repos/builtin/packages/esmf/package.py
@@ -90,6 +90,8 @@ class Esmf(MakefilePackage):
     )
     variant("debug", default=False, description="Make a debuggable version of the library")
     variant("shared", default=True, description="Build shared library")
+    variant("esmf_comm", default="auto", description="Override for ESMF_COMM variable")
+    variant("esmf_os", default="auto", description="Override for ESMF_OS variable")
 
     # Required dependencies
     depends_on("zlib")
@@ -251,6 +253,11 @@ class Esmf(MakefilePackage):
         if self.compiler.name == "cce" or "^cray-mpich" in self.spec:
             os.environ["ESMF_OS"] = "Unicos"
 
+        # Allow override of ESMF_OS:
+        os_variant = spec.variants["esmf_os"].value
+        if os_variant != "auto":
+            os.environ["ESMF_OS"] = os_variant
+
         #######
         # MPI #
         #######
@@ -283,6 +290,11 @@ class Esmf(MakefilePackage):
         else:
             # Force use of the single-processor MPI-bypass library.
             os.environ["ESMF_COMM"] = "mpiuni"
+
+        # Allow override of ESMF_COMM:
+        comm_variant = spec.variants["esmf_comm"].value
+        if comm_variant != "auto":
+            os.environ["ESMF_COMM"] = comm_variant
 
         ##########
         # LAPACK #

--- a/var/spack/repos/builtin/packages/esmf/package.py
+++ b/var/spack/repos/builtin/packages/esmf/package.py
@@ -90,6 +90,9 @@ class Esmf(MakefilePackage):
     )
     variant("debug", default=False, description="Make a debuggable version of the library")
     variant("shared", default=True, description="Build shared library")
+    # 'esmf_comm' and 'esmf_os' variants allow override values for their corresponding
+    # build environment variables. Documentation, including valid values, can be found at
+    # https://earthsystemmodeling.org/docs/release/latest/ESMF_usrdoc/node10.html#SECTION000105000000000000000
     variant("esmf_comm", default="auto", description="Override for ESMF_COMM variable")
     variant("esmf_os", default="auto", description="Override for ESMF_OS variable")
 

